### ISSUE TITLE
withTimeout missing a clearTimeout

### ISSUE THREAD

**Issue by cantoine** [DirtyHairy/async-mutex#33]
First off, thanks for the great library. 

While writing some unit tests on some code that uses Semaphores I noticed I was having Jest yell at me about workers not gracefully exiting. In trying to track down the issue with my tests I noticed it only happened if I was testing code that had a withTimeout and Semaphore or a Mutex. I did some digging and it appears that there is a missing `clearTimeout(handle)` when a lock is successfully acquired here: https://github.com/DirtyHairy/async-mutex/blob/master/src/withTimeout.ts#L26-L28 so timers are leaking on successful acquisition of locks.

I am opening a PR to add the clearTimeout and some tests. Please let me know of any changes you'd like me to make of if I've missed a reason for the lack of the clearTimeout.

Thanks!

**Comment by DirtyHairy** [comment]
Hi @cantoine !

The `clearTimeout` is not strictly necessary here as trying to reject an already resolved promise is a noop. Actually, the code didn't have a `clearTimeout` at all before I added support for `cancel`, it was only then that I added it in the catch handler (but more for reasons of tidyness than for anything else).

That said, if it is in the catch handler it should probably be in the main codepath as well, And Jest screaming because of pending timers is not ideal either 😏 So, thanks for your contribution, I'll review and merge it later.

Cheers
    -Christian
